// @dattabot-blockchain/geneth-tx/src/helpers/logParser.ts
export var attestLogParser = [{
    type: 'bytes32',
    name: 'encryptedBytes'
},{
    type: 'bytes32',
    name: 'version',
    indexed: true
},{
    type: 'address',
    name: 'itemAddress',
    indexed: true
},{
    type: 'bytes32',
    name: 'topic',
    indexed: true
},{
    type: 'address',
    name: 'attestor'
}];

export var whoAttestLogParser = [{
    type: 'bytes32',
    name: 'encryptedBytes'
},{
    type: 'bytes32',
    name: 'version'
},{
    type: 'address',
    name: 'itemAddress'
},{
    type: 'bytes32',
    name: 'topic'
},{
    type: 'address',
    name: 'attestor',
    indexed: true
}];

export var attestResultLogParser = [{
    type: 'bytes32',
    name: 'encryptedBytes'
},{
    type: 'bytes32',
    name: 'prevValue'
},{
    type: 'uint',
    name: 'prevExpiredTime'
},{
    type: 'bytes32',
    name: 'newValue'
},{
    type: 'uint',
    name: 'newExpiredTime'
}];

export var attestActionLogParser = [{
    type: 'bytes32',
    name: 'encryptedBytes'
},{
    type: 'bytes32',
    name: 'version'
},{
    type: 'address',
    name: 'itemAddress'
},{
    type: 'bytes32',
    name: 'topic'
},{
    type: 'address',
    name: 'attestor'
},{
    type: 'bytes32',
    name: 'value'
},{
    type: 'uint',
    name: 'expiredTime'
}];
